Redis: Sample upsert and bulk upsert endpoints

pallavi2209 opened the pull request.

iamigo commented on api/v1/controllers/samples.js:
```diff
@@ -159,25 +159,46 @@ module.exports = {
+        sampleQueryBody, resultObj, res.method
+      )
+      .then((sampleRes) => {
+        resultObj.dbTime = new Date() - resultObj.reqStartTime;
```
lines 170-180 and lines 189-200 are almost the same... I don't want you to spend more than a couple of minutes thinking about this, but if it's easy to do, can the duplicate-ish code be merged?

iamigo commented on api/v1/controllers/samples.js:
```diff
@@ -196,25 +217,43 @@ module.exports = {
+        redisModelSample.bulkUpsertSampleInRedis(value);
       }
-    });
+    } else {
```
Same comment here as above... if we're able to easily collapse the near duplicate code into one code block, it's less surface area to hide a bug and fewer lines to test and fewer lines of code to maintain

iamigo commented on cache/models/samples.js:
```diff
@@ -250,6 +267,142 @@ function applyFiltersOnSampObjs(sampObjArray, opts) {
+  const subjAspArr = sampleName.toLowerCase().split('|');
+  if (subjAspArr.length < TWO) {
+    throw new redisErrors.ResourceNotFoundError({
+      explanation: 'Incorrect sample name.',
```
If we're throwing ResourceNotFoundError the explanation should be something like 'no sample was found with this name' but here it feels more like an "Invalid..." error because a sample name without a pipe character can never be valid, can never be "found".

iamigo commented on cache/models/samples.js:
```diff
@@ -250,6 +267,142 @@ function applyFiltersOnSampObjs(sampObjArray, opts) {
+    sampleStore.toKey(constants.objectType.sample, sampleName),
+  ]);
+
+  return redisClient.batch(commands).execAsync()
```
I think here we will end up sending all three commands, but if we're throwing an error if the first comes back false, then why would we want to issue the other commands? And same for second?

iamigo commented on cache/models/samples.js:
```diff
@@ -375,4 +528,49 @@ module.exports = {
+   * @param  {String} method - Type of request method
+   * @returns {Promise} - Resolves to upserted sample
+   */
+  upsertSampleInRedis(qbObj, logObject, method) {
```
I think I prefer naming these functions *without* the "...InRedis" suffix--the fact that they're in *this* "cache" module and not the "db" module says that already, and without the suffix it sets up an equivalence/interchangeability between the db functions and the redis functions which I like

iamigo commented on cache/sampleStore.js:
```diff
@@ -68,7 +68,7 @@ function getNameFromKey(key) {
 function arrayStringsToJson(obj, arrayFields) {
   arrayFields.forEach((field) => {
-    if (obj[field] && !Array.isArray(obj[field])) {
+    if (obj && obj[field] && !Array.isArray(obj[field])) {
```
good catch, thx!

iamigo commented on cache/sampleStoreInit.js:
```diff
@@ -46,7 +47,7 @@ function eradicate() {
  */
 function populateAspects() {
-  return Aspect.findAll()
+  return Aspect.findAll({ where: { isPublished: true } })
```
thx!

iamigo commented on cache/sampleStoreInit.js:
```diff
@@ -63,6 +64,25 @@ function populateAspects() {
+    return redisClient.batch(cmds).execAsync();
+  })
+  .catch(console.error); // eslint-disable-line no-console
+} // populateAspects
```
// populateSubjects

iamigo commented on cache/sampleStoreInit.js:
```diff
@@ -137,7 +156,7 @@ function populate() {
   console.log(msg); // eslint-disable-line no-console
 
-  const promises = [populateSamples(), populateAspects()];
+  const promises = [populateSamples(), populateAspects(), populateSubjects()];
```
This doesn't matter at an implementation/execution level here, but I think subjects and aspects should be listed in this array *before* samples--it (very subtly) helps communicate that subjects and aspects have to exist for a sample to exist

pallavi2209 commented on cache/models/samples.js:
```diff
@@ -250,6 +267,142 @@ function applyFiltersOnSampObjs(sampObjArray, opts) {
+      props.push(sampFields.STS_CHNGED_AT, new Date().toString());
+      props.push(sampFields.STATUS, status);
+    }
+  }
```
@shriramshankar @iamigo Please take a look at this logic to confirm that we update/create sample fields correctly.

iamigo replied:
looks right to me

shriramshankar replied:
Any reason why you are using the props array? You can update the qbObj with the new fields and pass it to the hmset command.

pallavi2209 replied:
@shriramshankar No reason, just realized we can pass an object too. Lets do it then. Thanks! :)

shriramshankar commented on api/v1/controllers/samples.js:
```diff
@@ -196,25 +217,43 @@ module.exports = {
       } else {
-        helper.model.bulkUpsertByName(value,
-          userName);
+        redisModelSample.bulkUpsertSampleInRedis(value);
```
Can we change this to return the list of samples, to be used by the real-time publisher?

pallavi2209 commented:
@shriramshankar @iamigo Updated the PR.

pallavi2209 commented:
@iamigo @shriramshankar updated the PR. Some changes since you last reviewed:
1. Returning samples from bulk upsert method now, Shriram's comment
2. Separated redis batch commands based on Ian's comment
3. Removed props and updated queryObject to upsert, based on Shrirams comments
4. Added tests for bulk upsert job Queue
5. Some refactoring based on comments
